keps/0027-force-pod-reload-parameters.md

```diff
@@ -0,0 +1,203 @@
+      - ADDITIONAL_GROUP
+```
+
+KUDO then calculates the hash of all parameters in a group. This hash can then be used
+instead of the the `last-plan-execution-uid` in a deployment or stateful set to trigger the reload of the pods:
+
+```yaml
+spec:
+  template:
+    metadata:
+      annotations:
+        config-values-hash: {{ .ParamGroups.MAIN_GROUP.hash }}
+```
+
+When a parameter in the group changes, the hash will change, which in turn would then trigger the restart of the
+pods.
+
+This proposal would remove the `last-plan-execution-uid` and automatic pod restarts. Configuration for
+pod restarts would be fully in the responsibility of the operator developer.  
+
+It may be a good idea to define a `default` or `ALL` group that is always the set of all defined parameters.
+
+### User Stories
+
+- [#1424](https://github.com/kudobuilder/kudo/issues/1424) Fine control over pod restarts
+- [#1036](https://github.com/kudobuilder/kudo/issues/1036) Restarting all pods when scaling up
+
+The main use case for this extension are big stateful sets that are sensitive to pod restarts, for example Cassandra. 
+
+Especially while changing the size of a stateful set, the pod definition itself is not modified at all and a full
+restart of all pods will have a negative impact.
+
+### Implementation Details/Notes/Constraints
+
+The forced pod restart is currently implemented by setting a UID in the attributes of the `podTemplate`.
+
+Generally, the enhancer should not modify any template specs, but only the top-level resource.
+
+At the moment, the enhancer adds labels and attributes to templates:
+```
+	fieldsToAdd := map[string]string{
+		kudo.HeritageLabel: "kudo",
+		kudo.OperatorLabel: metadata.OperatorName,
+		kudo.InstanceLabel: metadata.InstanceName,
+	}
+	labelPaths := [][]string{
+		{"metadata", "labels"},
+		{"spec", "template", "metadata", "labels"},
+		{"spec", "volumeClaimTemplates[]", "metadata", "labels"},
+		{"spec", "jobTemplate", "metadata", "labels"},
+		{"spec", "jobTemplate", "spec", "template", "metadata", "labels"},
+	}
+
+```
+The labels are very static no change is required here. Labels can be used as selectors, and it still will be
+beneficial to have the heritage, operator and instance labels on templated pods.
+
+
+```
+	fieldsToAdd := map[string]string{
+		kudo.PlanAnnotation:            metadata.PlanName,
+		kudo.PhaseAnnotation:           metadata.PhaseName,
+		kudo.StepAnnotation:            metadata.StepName,
+		kudo.OperatorVersionAnnotation: metadata.OperatorVersion,
+		kudo.PlanUIDAnnotation:         string(metadata.PlanUID),
+	}
+	annotationPaths := [][]string{
+		{"metadata", "annotations"},
+		{"spec", "template", "metadata", "annotations"},
+		{"spec", "jobTemplate", "metadata", "annotations"},
+		{"spec", "jobTemplate", "spec", "template", "metadata", "annotations"},
+	}
+```
+The annotations are volatile and will/may change with every plan execution. The nested changes to template and jobTemplate
```

Comment:
I'm not necessarily advocating for keeping all annotations like you said they are modeled after kustomize. However, the only volatile annotation is `kudo.PlanUIDAnnotation`, right? The others may only if `OperatorVersion` changes.

Reply:
`OperatorVersion` would be safe to keep, yes. But Plan/Phase/Step might change, if a resource would be deployed by different plans ( one for deploy, one for a special update for example )

Reply:
You're right 👍